src/category_theory/lifting_properties.lean

```diff
@@ -0,0 +1,138 @@
+/-
+Copyright (c) 2021 Jakob Scholbach. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jakob Scholbach
+-/
+import category_theory.category
+import category_theory.arrow
+import category_theory.functor
+import category_theory.limits.preserves.shapes.terminal
+
+/-!
+# Lifting properties
+
+This file defines the lifting property of two arrows in a category and shows basic properties of
+this notion.
+We also construct the subcategory consisting of those morphisms which have the right lifting
+property with respect to arrows in a given diagram.
+
+## Main results
+- `has_lifting_property`: the definition of the lifting property
+- `iso_has_right_lifting_property`: any isomorphism satisfies the right lifting property (rlp)
+- `id_has_right_lifting_property`: any identity has the rlp
+- `right_lifting_property_initial_iff`: spells out the rlp with respect to a map whose source is an
+initial object
+- `right_lifting_subcat`: given a set of arrows `F : D → arrow C`, we construct the subcategory
+of those morphisms `p` in `C` that satisfy the rlp w.r.t. `F i`, for any element `i` of `D`.
+
+## Tags
+lifting property
+-/
+
+open category_theory.limits
+
+namespace category_theory
+
+universes v u v₁
+variables {C : Type u} [category.{v} C]
+variables {D : Type v₁}
+
+variables {X Y Z : C}
+
+/-- The lifting property of a morphism `i` with respect to a morphism `p`.
+This can be interpreted as the right lifting property of `i` with respect to `p`,
+or the left lifting property of `p` with respect to `i`. -/
+def has_lifting_property (i p : arrow C) : Prop := ∀ (sq : i ⟶ p), arrow.has_lift sq
+
+/-- Any isomorphism has the right lifting property with respect to any map.
+A    → X
+↓i    ↓p≅
+B    → Y
+-/
+lemma iso_has_right_lifting_property (i : arrow C) (p : X ≅ Y) :
+  has_lifting_property i (arrow.mk p.hom) :=
+λ sq, ⟨⟨{ lift := sq.right ≫ p.inv, }⟩⟩ -- the lift is obtained by p⁻¹ ∘ (B → Y)
+
+/-- Any identity has the right lifting property with respect to any map. -/
+lemma id_has_right_lifting_property (i : arrow C) : has_lifting_property i (arrow.mk (𝟙 X)) :=
+  iso_has_right_lifting_property i (iso.refl _)
+
+/-- An equivalent characterization for right lifting with respect to a map `i` whose source is
+initial.
+∅ → X
+↓   ↓
+B → Y has a lifting iff there is a map B → X making the right part commute.
+-/
+lemma right_lifting_property_initial_iff (i p : arrow C)
+  (h : is_initial i.left) :
+  has_lifting_property i p ↔ ∀ {e : i.right ⟶ p.right}, ∃ l : i.right ⟶ p.left, l ≫ p.hom = e :=
+begin
+  fsplit,
+  { intros hlift bottom,
+    have comm : (is_initial.to h p.left) ≫ p.hom = i.hom ≫ bottom :=
+      is_initial.hom_ext h _ _,
+    haveI := hlift (arrow.hom_mk comm),
+    use arrow.lift (arrow.hom_mk comm),
+    simp },
+  { intros h1 sq,
+    cases h1 with e he,
+    refine ⟨⟨{lift := e, fac_left' := _}⟩⟩,
+    apply is_initial.hom_ext, simpa using h },
```

Comment:
The proof is fine as is, but I you like you can shorten it to
```suggestion
    exact ⟨⟨⟨e, is_initial.hom_ext h _ _⟩⟩⟩
```

Reply:
Eventually this file will be very short :) Done!